experimental/packages/opentelemetry-sdk-metrics-base/src/export/InMemoryMetricExporter.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ExportResultCode } from "@opentelemetry/core"
+import { ExportResult } from "@opentelemetry/core"
+import { InstrumentType } from "../InstrumentDescriptor"
+import { AggregationTemporality } from "./AggregationTemporality"
+import { ResourceMetrics } from "./MetricData"
+import { PushMetricExporter } from "./MetricExporter"
+
+/**
+ * In-memory Metrics Exporter is a Push Metric Exporter
+ * which accumulates metrics data in the local memory and
+ * allows to inspect it (useful for e.g. unit tests).
+ */
+export class InMemoryMetricExporter implements PushMetricExporter {
+  protected _shutdown = false
+  protected _aggregationTemporality: AggregationTemporality
+  private _metrics: ResourceMetrics[] = []
+
+  constructor(aggregationTemporality: AggregationTemporality) {
+    this._aggregationTemporality = aggregationTemporality
+  }
+
+  /**
+   * @inheritedDoc
+   */
+  export(metrics: ResourceMetrics, resultCallback: (result: ExportResult) => void): void {
+    // Avoid storing metrics when exporter is shutdown
+    if (this. _shutdown) {
+      process.nextTick(() => resultCallback({ code: ExportResultCode.FAILED }));
+      return
+    }
+
+    this._metrics.push(metrics)
+    process.nextTick(() => resultCallback({ code: ExportResultCode.SUCCESS }));
+  }
+
+  /**
+   * Returns all the collected resource metrics
+   * @returns ResourceMetrics[]
+   */
+  public getMetrics(): ResourceMetrics[] {
+    return this._metrics
+  }
+
+  /**
+   * Clear the collected metrics
+   */
+  public clear(): void {
+    this._metrics = []
+  }
+
+  async forceFlush() {
```

Comment:
I think this should suffice as the method for clearing the metrics in memory mentioned [here](https://github.com/open-telemetry/opentelemetry-js/issues/3035#issuecomment-1155484368). :slightly_smiling_face: 
I'd say that in this case, we can remove `clear` again. :slightly_smiling_face:

Reply:
According to the spec, [`forceFlush`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#forceflush-1) stands for ensuring that the export of any Metrics the exporter has received. IMO, this means that for InMemoryExporter, it should be a no-op as the exporter pushes the received metrics to the buffer synchronously.

`forceFlush` can be called on conditions like teardown of the environment, like process exiting and document off-loading, e.g. https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/platform/browser/export/BatchSpanProcessor.ts#L33. It is true that if an exporter that has a buffer and a backend needs to "clear" the buffer and sent out everything it buffers. But this is not the case for InMemoryExporter, as its destination is this particular "buffer", and no backend to send. People can still inspect the "buffer" of InMemoryExporter after the SDK is shut down for postmortem purposes.

Reply:
How is this handled in the in memory span exporter? IMO we should remain consistent with ourselves

Reply:
There is a [`reset`](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/export/InMemorySpanExporter.ts#L54) method dedicated for resetting memory buffer. 

However, AFAICT, the signal of `forceFlush` in trace SDK is not passed to the span exporter. So the InMemorySpanExporter doesn't have a `forceFlush` method.

Reply:
Yeah, sounds good to me. I don't think it is meaningful to buffering with exporters while we already have a model of MetricReader which controls when to collect metrics.

Reply:
But I do find `forceFlush` is defined for SpanExporter in the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush-2

Reply:
possibly. I guess buffering in the exporter is discouraged but not impossible. JS is an odd duck because it's so hard to do async work during shutdown with any reliability unless you have full control over the shutdown process.

Reply:
What is the decision? What do you want me to change it to?

Reply:
Sorry about the ambiguity here. For SpanExporter, @dyladan has opened an issue to track the problem: https://github.com/open-telemetry/opentelemetry-js/issues/3067. For this particular `InMemoryMetricExporter.forceFlush`, I would prefer to follow what the spec is defined.

@pichlermarc since you are the initial requester for changing the behavior here, would you mind sharing your opinion again?

Reply:
You brought up a lot of good points :slightly_smiling_face:.
I think having `forceFlush()` be a no-op and having a separate `reset()` to reset the exported data makes more sense than removing `reset()`. 

Sorry for the back and forth I caused :sweat:

Reply:
I also think having an explicit `reset` function is a good idea. "flushing" doesn't really describe what the reset is doing in an obvious way, and maintaining symmetry with the span export API will be easier for users to learn